src/queue.ts

```diff
@@ -0,0 +1,1052 @@
+            ErrorCodes.Invalid_Input,
+            `the first argument 'input' is required by sendMessageBatch`
+          );
+        }
+
+        this.resource.grantSendMessages(context.role);
+
+        if (this.serializer.dataType === DataType.String) {
+          // when the data type is a String, no extra processing is required
+          // so we evaluate the entries to a JsonPath or Literal and pass it straight through
+          return context.evalExprToJsonPathOrLiteral(entries, (entries) => {
+            if (
+              ASLGraph.isJsonPath(entries) ||
+              (entries.value !== null && typeof entries.value === "object")
+            ) {
+              return context.stateWithHeapOutput({
+                Type: "Task",
+                Resource: "arn:aws:states:::aws-sdk:sqs:sendMessageBatch",
+                Parameters: {
+                  QueueUrl: this.queueUrl,
+                  ...(ASLGraph.isJsonPath(entries)
+                    ? {
+                        "Entries.$": `${entries.jsonPath}.Entries`,
+                      }
+                    : (entries.value as object)),
+                },
+                Next: ASLGraph.DeferNext,
+              });
+            } else {
+              throw new SynthError(
+                ErrorCodes.Invalid_Input,
+                `invalid 'Entries' property in SendMessageBatchRequest`
+              );
+            }
+          });
+        } else if (this.serializer.dataType === DataType.Json) {
+          // when the data type is Json, map over each of the entries and serialize them to JSON
+          return context.evalExprToJsonPath(
+            entries,
+            (entries, { addState }) => {
+              const heapVar = context.newHeapVariable();
+              addState({
+                Type: "Map",
+                ItemsPath: `${entries.jsonPath}.Entries`,
+                Parameters: {
+                  "entry.$": "$$.Map.Item.Value",
+                },
+                ResultPath: heapVar,
+                Next: ASLGraph.DeferNext,
+                Iterator: {
+                  StartAt: "serialize Message",
+                  States: {
+                    "serialize Message": {
+                      Type: "Pass",
+                      Parameters: {
+                        "value.$":
+                          this.serializer.dataType === DataType.Json
+                            ? `States.JsonToString($.entry.MessageBody)`
+                            : "$.entry.MessageBody",
+                      },
+                      ResultPath: "$.entry.MessageBody",
+                      Next: "unwrap Message",
+                    },
+                    "unwrap Message": {
+                      Type: "Pass",
+                      InputPath: "$.entry.MessageBody.value",
+                      ResultPath: "$.entry.MessageBody",
+                      OutputPath: "$.entry",
+                      End: true,
+                    },
+                  },
```

Comment:
you should use `context.aslGraphToStates`, which will ensure the state names are globally unique, apply optimization, and you can use ASLGraph.

https://github.com/functionless/functionless/blob/main/src/step-function.ts#L456

Reply:
Why is that the callee's responsibility

Reply:
Just is right now. We could process map, parallel, etc at the end to ensure any January created state names are unique with sub-graphs.

Reply:
Does it matter if I ignore the return value of `aslGraphToStates`?

Reply:
I thought that we could just chain together states and the underlying machine would flatten them out into a machine. That doesn't seem to be the case?

Reply:
The types suggest I can return `startsAt` and `states` SubState from evalExpr?

<img width="537" alt="image" src="https://user-images.githubusercontent.com/38672686/188706214-bdc38fcb-2a44-4cf6-b25d-648cdb1d0176.png">

Reply:
Should I use `jsonPath` or `output: { jsonPath } ` for the heap var?

Reply:
Got it working with this code! I assume your suggestion of `aslGraphToStates` was wrong?

```ts
return context.evalExprToJsonPath(entries, (entries) => {
  const heapVar = context.newHeapVariable();

  return {
    output: {
      jsonPath: heapVar,
    },
    startState: "serialize messages",
    states: {
      "serialize messages": {
        Type: "Map",
        ItemsPath: `${entries.jsonPath}.Entries`,
        Parameters: {
          "entry.$": "$$.Map.Item.Value",
        },
        ResultPath: heapVar,
        Next: "send message batch",
        Iterator: {
          StartAt: "serialize Message",
          States: {
            "serialize Message": {
              Type: "Pass",
              Parameters: {
                "value.$": `States.JsonToString($.entry.MessageBody)`,
              },
              ResultPath: "$.entry.MessageBody",
              Next: "unwrap Message",
            },
            "unwrap Message": {
              Type: "Pass",
              InputPath: "$.entry.MessageBody.value",
              ResultPath: "$.entry.MessageBody",
              OutputPath: "$.entry",
              End: true,
            },
          },
        },
      },
      "send message batch": {
        Type: "Task",
        Resource: "arn:aws:states:::aws-sdk:sqs:sendMessageBatch",
        Parameters: {
          QueueUrl: this.queueUrl,
          "Entries.$": heapVar,
        },
        ResultPath: heapVar,
        Next: ASLGraph.DeferNext,
      },
    },
  };
});
```

Reply:
Nope, my comment was just about `Iterator`. `aslGraphToStates` is a terminal call in the chain that output the final ASL format. The contents of `Map.Iterator` and `Parallel.Branches[]` are complete ASL graphs (`{StartAt, States}`) (which I have been calling SubGraphs). 

The suggestion was to do:

```ts
return context.evalExprToJsonPath(entries, (entries) => {
  const heapVar = context.newHeapVariable();

  return {
    output: {
      jsonPath: heapVar,
    },
    startState: "serialize messages",
    states: {
      "serialize messages": {
        Type: "Map",
        ItemsPath: `${entries.jsonPath}.Entries`,
        Parameters: {
          "entry.$": "$$.Map.Item.Value",
        },
        ResultPath: heapVar,
        Next: "send message batch",
        Iterator: context.aslGraphToStates({
          startState: "serialize Message",
          states: {
            "serialize Message": {
              Type: "Pass",
              Parameters: {
                "value.$": `States.JsonToString($.entry.MessageBody)`,
              },
              ResultPath: "$.entry.MessageBody",
              Next: "unwrap Message",
            },
            "unwrap Message": {
              Type: "Pass",
              InputPath: "$.entry.MessageBody.value",
              ResultPath: "$.entry.MessageBody",
              OutputPath: "$.entry",
              End: true,
            },
          }),
        },
      },
      "send message batch": {
        Type: "Task",
        Resource: "arn:aws:states:::aws-sdk:sqs:sendMessageBatch",
        Parameters: {
          QueueUrl: this.queueUrl,
          "Entries.$": heapVar,
        },
        ResultPath: heapVar,
        Next: ASLGraph.DeferNext,
      },
    },
  };
});
```

Yours will work, I think, util someone uses batch twice in the same machine or managed to create any other state with the same names.

Reply:
The contents of Iterstor is an isolated machine right?